### PR TITLE
Trying to make sense of the streaming code

### PR DESCRIPTION
**Pull request by Bond-009**
Mostly small changes as I was looking through the code.

* async void -> async Task
* Properly implemented dispose methods
* Pass the logstream directly to the JobLogger
* Style fixes

### CODE REVIEW

File: MediaBrowser.Api/Playback/BaseStreamingService.cs [hunk 1]
```diff
@@ -133,31 +137,18 @@ protected virtual string GetOutputFileExtension(StreamState state)
...
             var data = GetCommandLineArguments("dummy\\dummy", encodingOptions, state, false);
 
-            data += "-" + (state.Request.DeviceId ?? string.Empty);
-            data += "-" + (state.Request.PlaySessionId ?? string.Empty);
-
-            var dataHash = data.GetMD5().ToString("N");
+            data += "-" + (state.Request.DeviceId ?? string.Empty)
+                 + "-" + (state.Request.PlaySessionId ?? string.Empty);
 
-            if (EnableOutputInSubFolder)
```
Comment (JustAMan) [inline MediaBrowser.Api/Playback/BaseStreamingService.cs]: why drop this `if`?

File: MediaBrowser.Api/Playback/BaseStreamingService.cs [hunk 2]
```diff
@@ -261,27 +245,26 @@ private async Task AcquireResources(StreamState state, CancellationTokenSource c
...
                 }
-                else if (string.Equals(state.OutputVideoCodec, "copy", StringComparison.OrdinalIgnoreCase))
+                else
                 {
                     logFilePrefix = "ffmpeg-remux";
                 }
             }
 
             var logFilePath = Path.Combine(ServerConfigurationManager.ApplicationPaths.LogDirectoryPath, logFilePrefix + "-" + Guid.NewGuid() + ".txt");
-            Directory.CreateDirectory(Path.GetDirectoryName(logFilePath));
```
Comment (JustAMan) [inline MediaBrowser.Api/Playback/BaseStreamingService.cs]: why remove?

File: MediaBrowser.Api/Playback/BaseStreamingService.cs [hunk 3]
```diff
@@ -368,25 +348,16 @@ private void OnFfMpegProcessExited(Process process, TranscodingJob job, StreamSt
...
+                Logger.LogInformation("FFMpeg exited with code 0");
             }
-            catch
+            else
             {
-                Logger.LogError("FFMpeg exited with an error.");
+                Logger.LogError("FFMpeg exited with code {0}", process.ExitCode);
             }
 
-            // This causes on exited to be called twice:
```
Comment (JustAMan) [inline MediaBrowser.Api/Playback/BaseStreamingService.cs]: what was the issue that was causing double call, and why it is no longer an issue here?

File: MediaBrowser.Api/Playback/BaseStreamingService.cs [hunk 4]
```diff
@@ -643,11 +614,19 @@ private void ParseDlnaHeaders(StreamRequest request)
...
             }
-            value = value.Substring(4).Split(new[] { '-' }, 2)[0];
+            int index = value.IndexOf('-');
+            if (index == -1)
+            {
+                value = value.Substring(4);
+            }
+            else
+            {
+                value = value.Substring(4, index);
```
Comment (JustAMan) [inline MediaBrowser.Api/Playback/BaseStreamingService.cs]: huh, what this does? I'm not sure it is equivalent to the older code.
Also what is the reason for the change?

File: MediaBrowser.Api/Playback/BaseStreamingService.cs [hunk 5]
```diff
@@ -779,12 +755,12 @@ protected async Task<StreamState> GetState(StreamRequest request, CancellationTo
                     var mediaSources = (await MediaSourceManager.GetPlayackMediaSources(LibraryManager.GetItemById(request.Id), null, false, false, cancellationToken).ConfigureAwait(false)).ToList();
 
                     mediaSource = string.IsNullOrEmpty(request.MediaSourceId)
-                       ? mediaSources.First()
-                       : mediaSources.FirstOrDefault(i => string.Equals(i.Id, request.MediaSourceId));
+                       ? mediaSources[0]
+                       : mediaSources.Find(i => string.Equals(i.Id, request.MediaSourceId));
```
Comment (JustAMan) [inline MediaBrowser.Api/Playback/BaseStreamingService.cs]: why replace `FirstOrDefault` with `Find`?

File: MediaBrowser.Controller/MediaEncoding/JobLogger.cs [hunk 6]
```diff
@@ -97,8 +99,7 @@ private void ParseLogLine(string line, EncodingJobInfo state)
                     {
                         var currentMs = startMs + val.TotalMilliseconds;
 
-                        var percentVal = currentMs / totalMs;
-                        percent = 100 * percentVal;
+                        percent = 100 * currentMs / totalMs;
```
Comment (JustAMan) [inline MediaBrowser.Controller/MediaEncoding/JobLogger.cs]:
```suggestion
                        percent = 100.0 * currentMs / totalMs;
```
for clarity that we're getting a floating point here... and to really get a floating point if both `currentMs` and `totalMs` are integers :)

File: MediaBrowser.Api/Playback/BaseStreamingService.cs (hunk 2 above)
Comment (Bond-009) [reply MediaBrowser.Api/Playback/BaseStreamingService.cs]: It's the log dir, there would be a big problem if it didn't exist.

File: MediaBrowser.Api/Playback/BaseStreamingService.cs (hunk 1 above)
Comment (Bond-009) [reply MediaBrowser.Api/Playback/BaseStreamingService.cs]: Woops

File: MediaBrowser.Api/Playback/BaseStreamingService.cs (hunk 3 above)
Comment (Bond-009) [reply MediaBrowser.Api/Playback/BaseStreamingService.cs]: The base object has changed from a custom one to the std one and disposing multiple times shouldn't cause an issue.

File: MediaBrowser.Api/Playback/BaseStreamingService.cs (hunk 4 above)
Comment (Bond-009) [reply MediaBrowser.Api/Playback/BaseStreamingService.cs]: This strips the "npt=" from the beginning of the string and everything after '-'.
I changed it because `Split` is kind of hacky for this (at least imo)

File: MediaBrowser.Api/Playback/BaseStreamingService.cs (hunk 5 above)
Comment (Bond-009) [reply MediaBrowser.Api/Playback/BaseStreamingService.cs]: `FirstOrDefault` is a generic function from LINQ.
`Find` is a function specifically for `Lists`.
In other words: `Find` is faster 🤷‍♂️

File: MediaBrowser.Api/Playback/BaseStreamingService.cs (hunk 4 above)
Comment (JustAMan) [reply MediaBrowser.Api/Playback/BaseStreamingService.cs]: Add a comment explaining that `4 == len("npt=")` please :)

Comment (Bond-009) [reply MediaBrowser.Api/Playback/BaseStreamingService.cs]: Made it self-documenting ;)

File: MediaBrowser.Api/Playback/BaseStreamingService.cs [hunk 7]
```diff
@@ -32,6 +30,10 @@ namespace MediaBrowser.Api.Playback
     /// </summary>
     public abstract class BaseStreamingService : BaseApiService
     {
+        protected static readonly CultureInfo UsCulture = CultureInfo.ReadOnly(new CultureInfo("en-US"));
```
Comment (EraYaN) [inline MediaBrowser.Api/Playback/BaseStreamingService.cs]: UsCulture variable is only used for int and double parsing and one formatting call. We should use `CultureInfo.InvariantCulture` I believe.
